dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+                                options=self.options)
+            else:
+                opt = Optimizer(self.metric.value_and_gradient, self.x0,
+                                method=self.method, jac=True,
+                                options=self.options)
+            xopt = opt.xopt
+
+            # Update prealign matrix with optimal parameters
+            T = self.transform.param_to_matrix(xopt)
+            self.prealign = T.dot(self.prealign)
+
+            # Start next iteration at identity
+            self.x0 = self.transform.get_identity_parameters()
+
+            # Update the metric to the current solution
+            self.metric._update_dense(xopt, False)
+        return self.prealign
+
+
+def aff_warp(static, static_grid2space, moving, moving_grid2space, transform,
+             nn=False):
+    r""" Warps the moving image towards the static using the given transform
+
+    Parameters
+    ----------
+    static: array, shape(S, R, C)
+        static image: it will provide the grid and grid-to-space transform for
+        the warped image
+    static_grid2space:
+        grid-to-space transform associated with the static image
+    moving: array, shape(S', R', C')
+        moving image
+    moving_grid2space:
+        grid-to-space transform associated with the moving image
+    transform: array, shape (dim+1, dim+1)
+        the matrix representing the affine transform to be applied to `moving`
+    nn : Boolean, optional
+        if False, trilinear interpolation will be used. If True, nearest
+        neighbour interpolation will be used instead. The default is False,
+        implying trilinear interpolation.
+    Returns
+    -------
+    warped: array, shape (S, R, C)
+        the warped image
+    """
+    if type(static) is tuple:
+        dim = len(static)
+        shape = np.array(static, dtype=np.int32)
+    else:
+        dim = len(static.shape)
+        shape = np.array(static.shape, dtype=np.int32)
+    if nn:
+        input = np.array(moving, dtype=np.int32)
+        if dim == 2:
+            warp_method = vf.warp_2d_affine_nn
+        elif dim == 3:
+            warp_method = vf.warp_3d_affine_nn
+    else:
+        input = np.array(moving, dtype=floating)
+        if dim == 2:
+            warp_method = vf.warp_2d_affine
+        elif dim == 3:
+            warp_method = vf.warp_3d_affine
+    if moving_grid2space is not None:
+        m_space2grid = npl.inv(moving_grid2space)
+    else:
+        m_space2grid = np.eye(dim + 1)
+
+    if static_grid2space is None:
+        static_grid2space = np.eye(dim + 1)
+
+    if transform is None:
+        composition = m_space2grid.dot(static_grid2space)
+    else:
+        composition = m_space2grid.dot(transform.dot(static_grid2space))
+
+    warped = warp_method(input, shape, composition)
+
+    return np.array(warped)
+
+
+def aff_centers_of_mass(static, static_grid2space, moving, moving_grid2space):
```

Comment:
You are in the imaffine.py file. There is no need to start all functions with aff.